File: test/clients/go/routines_test.go
// Routine + view lifecycle:
//   - SQL UDFs (CREATE FUNCTION, call, list via routines REST)
//   - Procedures with IN-params used inside INSERT bodies
//   - Procedures whose body is a SELECT — rows propagate to CALL
//   - Views
//   - Materialized views + CALL BQ.REFRESH_MATERIALIZED_VIEW
package bqlocal_test

import (
	"context"
	"testing"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
)

func seedDataset(t *testing.T, client *bigquery.Client) {
	t.Helper()
	if err := client.Dataset("ds").Create(context.Background(), &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
}

func TestSQLUDFRoundTrip(t *testing.T) {
	client, project := newClient(t)
	seedDataset(t, client)
	if _, err := readAll(t, client.Query(
		"CREATE FUNCTION `"+project+".ds.double`(x INT64) RETURNS INT64 AS (x * 2)")); err != nil {
		t.Fatalf("create udf: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT `"+project+".ds.double`(21) AS n"))
	if err != nil {
		t.Fatalf("call udf: %v", err)
	}
	if rows[0][0].(int64) != 42 {
		t.Errorf("double(21) = %v, want 42", rows[0][0])
	}
}

func TestProcedureWithInParamInInsertBody(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	seedDataset(t, client)
	table := client.Dataset("ds").Table("audit")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "name", Type: bigquery.StringFieldType},
			{Name: "ts", Type: bigquery.TimestampFieldType},
		},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}
	if _, err := readAll(t, client.Query(`
		CREATE PROCEDURE `+"`"+project+".ds.record_visit`"+`(IN name STRING)
		BEGIN
		  INSERT INTO `+"`"+project+".ds.audit`"+` (name, ts) VALUES (name, CURRENT_TIMESTAMP());
		END;
	`)); err != nil {
		t.Fatalf("create procedure: %v", err)
	}
	for _, n := range []string{"alice", "bob"} {
		if _, err := readAll(t, client.Query(
			"CALL `"+project+".ds.record_visit`('"+n+"')")); err != nil {
			t.Fatalf("call %s: %v", n, err)
		}
	}
	rows, _ := readAll(t, client.Query(
		"SELECT name FROM `"+project+".ds.audit` ORDER BY name"))
	if len(rows) != 2 || rows[0][0].(string) != "alice" || rows[1][0].(string) != "bob" {
		t.Errorf("audit rows = %v, want [alice bob]", rows)
	}
}

func TestProcedureReturningSelectSurfacesRows(t *testing.T) {
	client, project := newClient(t)
	seedDataset(t, client)
	if _, err := readAll(t, client.Query(`
		CREATE PROCEDURE `+"`"+project+".ds.greet`"+`(IN who STRING)
		BEGIN
		  SELECT CONCAT('hi ', who) AS message;
		END;
	`)); err != nil {
		t.Fatalf("create procedure: %v", err)
	}
	rows, err := readAll(t, client.Query("CALL `"+project+".ds.greet`('alice')"))
	if err != nil {
		t.Fatalf("call: %v", err)
	}
	if len(rows) == 0 {
		t.Fatal("CALL returned no rows (BL-procedure SELECT body fix)")
	}
	if rows[0][0].(string) != "hi alice" {
		t.Errorf("message = %v, want hi alice", rows[0][0])
	}
}

func TestViewLifecycle(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	seedDataset(t, client)
	table := client.Dataset("ds").Table("orders")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "region", Type: bigquery.StringFieldType},
			{Name: "amount", Type: bigquery.IntegerFieldType},
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	type regionAmt struct {
		Region string `bigquery:"region"`
		Amount int64  `bigquery:"amount"`
	}
	if err := table.Inserter().Put(ctx, []*regionAmt{
		{Region: "east", Amount: 10},
		{Region: "east", Amount: 20},
		{Region: "west", Amount: 30},
	}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	if _, err := readAll(t, client.Query(
		"CREATE VIEW `"+project+".ds.east_orders` AS "+
			"SELECT region, amount FROM `"+project+".ds.orders` WHERE region = 'east'")); err != nil {
		t.Fatalf("create view: %v", err)
	}
	rows, _ := readAll(t, client.Query(
		"SELECT count(*)::INT64 AS n FROM `"+project+".ds.east_orders`"))
	if rows[0][0].(int64) != 2 {
		t.Errorf("view count = %v, want 2", rows[0][0])
	}
}

func TestMaterializedViewCreateAndRefresh(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	seedDataset(t, client)
	table := client.Dataset("ds").Table("orders")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "region", Type: bigquery.StringFieldType},
			{Name: "amount", Type: bigquery.IntegerFieldType},
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	type regionAmt struct {
		Region string `bigquery:"region"`
		Amount int64  `bigquery:"amount"`
	}
	if err := table.Inserter().Put(ctx, []*regionAmt{
		{Region: "east", Amount: 1},
		{Region: "east", Amount: 2},
	}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	if _, err := readAll(t, client.Query(
		"CREATE MATERIALIZED VIEW `"+project+".ds.east_total` AS "+
			"SELECT SUM(amount) AS total FROM `"+project+".ds.orders` WHERE region = 'east'")); err != nil {
		t.Fatalf("create mv: %v", err)
	}
	before, _ := readAll(t, client.Query("SELECT total FROM `"+project+".ds.east_total`"))
	if before[0][0].(int64) != 3 {
		t.Errorf("snapshot total = %v, want 3", before[0][0])
	}
	// Add a row — MV is stale.
	if err := table.Inserter().Put(ctx, []*regionAmt{{Region: "east", Amount: 10}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	stale, _ := readAll(t, client.Query("SELECT total FROM `"+project+".ds.east_total`"))
	if stale[0][0].(int64) != 3 {
		t.Errorf("stale total = %v, want still 3", stale[0][0])
	}
	// Refresh.
	if _, err := readAll(t, client.Query(
		"CALL BQ.REFRESH_MATERIALIZED_VIEW('" + project + ".ds.east_total')")); err != nil {
		t.Fatalf("refresh: %v", err)
	}
	refreshed, _ := readAll(t, client.Query("SELECT total FROM `"+project+".ds.east_total`"))
	if refreshed[0][0].(int64) != 13 {
		t.Errorf("refreshed total = %v, want 13", refreshed[0][0])
	}
}

func TestRoutinesRestGetAndList(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	seedDataset(t, client)
	if _, err := readAll(t, client.Query(
		"CREATE FUNCTION `"+project+".ds.tripler`(x INT64) RETURNS INT64 AS (x * 3)")); err != nil {
		t.Fatalf("create fn: %v", err)
	}
	it := client.Dataset("ds").Routines(ctx)
	var names []string
	for {
		r, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			t.Fatalf("routines list: %v", err)
		}
		names = append(names, r.RoutineID)
	}
	if len(names) != 1 || names[0] != "tripler" {
		t.Errorf("routines = %v, want [tripler]", names)
	}
	fetched, err := client.Dataset("ds").Routine("tripler").Metadata(ctx)
	if err != nil {
		t.Fatalf("get routine: %v", err)
	}
	if fetched.Body != "x * 3" {
		t.Errorf("body = %q, want %q", fetched.Body, "x * 3")
	}
}

File: test/clients/java/src/test/java/local/bigquery/RoutinesTest.java
package local.bigquery;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.DatasetId;
import com.google.cloud.bigquery.DatasetInfo;
import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.Routine;
import com.google.cloud.bigquery.RoutineId;
import com.google.cloud.bigquery.Schema;
import com.google.cloud.bigquery.StandardSQLTypeName;
import com.google.cloud.bigquery.StandardTableDefinition;
import com.google.cloud.bigquery.TableId;
import com.google.cloud.bigquery.TableInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** SQL UDFs, procedures, views, materialized views — mirrors routines_test.go. */
class RoutinesTest {
  @Test
  void sqlUdfRoundTrip() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    Emu.rows(bq, "CREATE FUNCTION `" + project + ".ds.double`(x INT64) RETURNS INT64 AS (x * 2)");
    List<FieldValueList> rows = Emu.rows(bq, "SELECT `" + project + ".ds.double`(21) AS n");
    assertEquals(42L, rows.get(0).get("n").getLongValue());
  }

  @Test
  void procedureWithInParamInInsertBody() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    bq.create(
        TableInfo.of(
            TableId.of("ds", "audit"),
            StandardTableDefinition.of(
                Schema.of(
                    Field.of("name", StandardSQLTypeName.STRING),
                    Field.of("ts", StandardSQLTypeName.TIMESTAMP)))));
    Emu.rows(
        bq,
        "CREATE PROCEDURE `"
            + project
            + ".ds.record_visit`(IN name STRING) BEGIN "
            + "INSERT INTO `"
            + project
            + ".ds.audit` (name, ts) VALUES (name, CURRENT_TIMESTAMP()); END;");
    for (String n : List.of("alice", "bob")) {
      Emu.rows(bq, "CALL `" + project + ".ds.record_visit`('" + n + "')");
    }
    List<FieldValueList> rows =
        Emu.rows(bq, "SELECT name FROM `" + project + ".ds.audit` ORDER BY name");
    assertEquals(2, rows.size());
    assertEquals("alice", rows.get(0).get("name").getStringValue());
    assertEquals("bob", rows.get(1).get("name").getStringValue());
  }

  @Test
  void procedureReturningSelectSurfacesRows() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    Emu.rows(
        bq,
        "CREATE PROCEDURE `"
            + project
            + ".ds.greet`(IN who STRING) BEGIN SELECT CONCAT('hi ', who) AS message; END;");
    List<FieldValueList> rows = Emu.rows(bq, "CALL `" + project + ".ds.greet`('alice')");
    assertEquals("hi alice", rows.get(0).get("message").getStringValue());
  }

  @Test
  void viewLifecycle() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId orders = TableId.of("ds", "orders");
    bq.create(
        TableInfo.of(
            orders,
            StandardTableDefinition.of(
                Schema.of(
                    Field.of("region", StandardSQLTypeName.STRING),
                    Field.of("amount", StandardSQLTypeName.INT64)))));
    bq.insertAll(
        InsertAllRequest.newBuilder(orders)
            .addRow(Map.of("region", "east", "amount", 10))
            .addRow(Map.of("region", "east", "amount", 20))
            .addRow(Map.of("region", "west", "amount", 30))
            .build());
    Emu.rows(
        bq,
        "CREATE VIEW `"
            + project
            + ".ds.east_orders` AS SELECT region, amount FROM `"
            + project
            + ".ds.orders` WHERE region = 'east'");
    List<FieldValueList> rows =
        Emu.rows(bq, "SELECT count(*) AS n FROM `" + project + ".ds.east_orders`");
    assertEquals(2L, rows.get(0).get("n").getLongValue());
  }

  @Test
  void materializedViewCreateAndRefresh() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId orders = TableId.of("ds", "orders");
    bq.create(
        TableInfo.of(
            orders,
            StandardTableDefinition.of(
                Schema.of(
                    Field.of("region", StandardSQLTypeName.STRING),
                    Field.of("amount", StandardSQLTypeName.INT64)))));
    bq.insertAll(
        InsertAllRequest.newBuilder(orders)
            .addRow(Map.of("region", "east", "amount", 1))
            .addRow(Map.of("region", "east", "amount", 2))
            .build());
    Emu.rows(
        bq,
        "CREATE MATERIALIZED VIEW `"
            + project
            + ".ds.east_total` AS SELECT SUM(amount) AS total FROM `"
            + project
            + ".ds.orders` WHERE region = 'east'");

    assertEquals(
        3L,
        Emu.rows(bq, "SELECT total FROM `" + project + ".ds.east_total`")
            .get(0)
            .get("total")
            .getLongValue());

    bq.insertAll(
        InsertAllRequest.newBuilder(orders).addRow(Map.of("region", "east", "amount", 10)).build());
    assertEquals(
        3L,
        Emu.rows(bq, "SELECT total FROM `" + project + ".ds.east_total`")
            .get(0)
            .get("total")
            .getLongValue());

    Emu.rows(bq, "CALL BQ.REFRESH_MATERIALIZED_VIEW('" + project + ".ds.east_total')");
    assertEquals(
        13L,
        Emu.rows(bq, "SELECT total FROM `" + project + ".ds.east_total`")
            .get(0)
            .get("total")
            .getLongValue());
  }

  @Test
  void routinesRestGetAndList() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    Emu.rows(bq, "CREATE FUNCTION `" + project + ".ds.tripler`(x INT64) RETURNS INT64 AS (x * 3)");

    List<String> names = new ArrayList<>();
    bq.listRoutines(DatasetId.of(project, "ds"))
        .iterateAll()
        .forEach(r -> names.add(r.getRoutineId().getRoutine()));
    assertEquals(List.of("tripler"), names);

    Routine fetched = bq.getRoutine(RoutineId.of(project, "ds", "tripler"));
    assertEquals("x * 3", fetched.getBody());
  }
}
